File: cogs/player_progress.py
import discord
from discord import app_commands
from discord.ext import commands, tasks

import db


class PlayerProgress(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        self.cleanup_loop.start()

    def cog_unload(self):
        self.cleanup_loop.cancel()

    # =========================
    # REGISTRAR LEVEL
    # =========================
    @app_commands.command(
        name="registrarlevel",
        description="Registrar level diário do jogador",
    )
    async def registrarlevel(
        self,
        interaction: discord.Interaction,
        jogador: discord.Member,
        level: int,
        dias_atras: int = 0,
    ):
        if dias_atras < 0 or dias_atras > 3:
            await interaction.response.send_message(
                "❌ Só é permitido até 3 dias retroativos.",
                ephemeral=True,
            )
            return

        try:
            db.add_player_level(jogador.id, jogador.display_name, level, dias_atras)
        except Exception:
            await interaction.response.send_message(
                "⚠️ Esse jogador já tem registro nesse dia.",
                ephemeral=True,
            )
            return

        await interaction.response.send_message(
            f"✅ Level **{level}** registrado para {jogador.mention}.",
            ephemeral=True,
        )

    # =========================
    # CHECAR QUEM ESTÁ PARADO
    # =========================
    @app_commands.command(
        name="checarniveis",
        description="Ver jogadores sem progresso nos últimos 3 dias",
    )
    async def checarniveis(self, interaction: discord.Interaction):
        stuck = db.get_players_stuck_3_days()

        if not stuck:
            await interaction.response.send_message(
                "🚀 Todos estão progredindo!",
                ephemeral=True,
            )
            return

        msg = "⚠️ **Jogadores sem progresso há 3 dias:**\n\n"
        for pid, name in stuck:
            msg += f"• <@{pid}> ({name})\n"

        await interaction.response.send_message(msg, ephemeral=True)

    # =========================
    # LIMPEZA AUTOMÁTICA
    # =========================
    @tasks.loop(hours=24)
    async def cleanup_loop(self):
        db.cleanup_old_players()


async def setup(bot):
    await bot.add_cog(PlayerProgress(bot))

File: node/cogs/player_progress.js
const { SlashCommandBuilder, PermissionFlagsBits } = require('discord.js');

const registerLevelCommand = new SlashCommandBuilder()
  .setName('registrarlevel')
  .setDescription('Registrar level diário do jogador')
  .addUserOption((option) =>
    option.setName('jogador').setDescription('Jogador').setRequired(true),
  )
  .addIntegerOption((option) =>
    option.setName('level').setDescription('Level').setRequired(true),
  )
  .addIntegerOption((option) =>
    option
      .setName('dias_atras')
      .setDescription('Dias retroativos')
      .setRequired(false),
  )
  .setDefaultMemberPermissions(PermissionFlagsBits.ManageMessages);

const checkLevelsCommand = new SlashCommandBuilder()
  .setName('checarniveis')
  .setDescription('Ver jogadores sem progresso nos últimos 3 dias')
  .setDefaultMemberPermissions(PermissionFlagsBits.ManageMessages);

async function handleRegisterLevel({ interaction, db }) {
  const jogador = interaction.options.getUser('jogador', true);
  const level = interaction.options.getInteger('level', true);
  const diasAtras = interaction.options.getInteger('dias_atras') || 0;

  if (diasAtras < 0 || diasAtras > 3) {
    await interaction.reply({
      content: '❌ Só é permitido até 3 dias retroativos.',
      ephemeral: true,
    });
    return;
  }

  try {
    await db.addPlayerLevel(jogador.id, jogador.username, level, diasAtras);
  } catch (error) {
    await interaction.reply({
      content: '⚠️ Esse jogador já tem registro nesse dia.',
      ephemeral: true,
    });
    return;
  }

  await interaction.reply({
    content: `✅ Level **${level}** registrado para ${jogador.toString()}.`,
    ephemeral: true,
  });
}

async function handleCheckLevels({ interaction, db }) {
  const stuck = await db.getPlayersStuck3Days();
  if (!stuck || stuck.length === 0) {
    await interaction.reply({ content: '🚀 Todos estão progredindo!', ephemeral: true });
    return;
  }

  const lines = stuck.map((row) => `• <@${row.player_id}> (${row.player_name})`);
  const message = `⚠️ **Jogadores sem progresso há 3 dias:**\n\n${lines.join('\n')}`;
  await interaction.reply({ content: message, ephemeral: true });
}

function startCleanupLoop({ db }) {
  const interval = setInterval(async () => {
    await db.cleanupOldPlayers();
  }, 24 * 60 * 60 * 1000);

  return () => clearInterval(interval);
}

function registerPlayerProgress({ client, db }) {
  startCleanupLoop({ db });

  client.on('interactionCreate', async (interaction) => {
    if (!interaction.isChatInputCommand()) {
      return;
    }

    if (interaction.commandName === 'registrarlevel') {
      await handleRegisterLevel({ interaction, db });
      return;
    }

    if (interaction.commandName === 'checarniveis') {
      await handleCheckLevels({ interaction, db });
    }
  });
}

module.exports = {
  commands: [registerLevelCommand, checkLevelsCommand],
  registerPlayerProgress,
};
